### ISSUE TITLE
Quarkus debug: Cannot connect - Failed to handshake

### ISSUE THREAD

**Issue by jcribeirooo** [mfussenegger/nvim-dap#451]
### Debug adapter definition and debug configuration

:lua require'dap'.run({ type = 'java'; request = 'attach'; name = "Debug (Attach) - Remote"; hostName = "127.0.0.1"; port = 5005; })

### Debug adapter version

Quarkus (2.7.1.Final) - built-in debugger

### Steps to Reproduce

1. Create a new Java Quarkus project (https://quarkus.io/);
2. Execute "mvn quarkus:dev -Dsuspend" to wait for debugger to be attached
3. Execute in lvim: ':lua require'dap'.run({ type = 'java'; request = 'attach'; name = "Debug (Attach) - Remote"; hostName = "127.0.0.1"; port = 5005; })"

### Expected Result

To be able to attach the debugger.

### Actual Result

Listening for transport dt_socket at address: 5005
Debugger failed to attach: handshake failed - received >Content-Length< - expected >JDWP-Handshake<

**Comment by mfussenegger** [comment]
Can you also show your `dap.adapters.java` definition?

**Comment by jcribeirooo** [comment]
I tried with and without this config (dap.adapters.java). Either way it gives me the same error.

local dap = require('dap')
dap.adapters.java = function(callback)
  callback({
    type = 'server';
    host = '127.0.0.1';
    port = 5005;
  })
end

**Comment by mfussenegger** [comment]
I suggest to read the docs: https://github.com/mfussenegger/nvim-dap/wiki/Java

**Comment by jcribeirooo** [comment]
I could get it working! Thank you for pointing it out.

**Comment by wviana** [comment]
@jcribeirooo please share what was it. I'm getting same here.

**Comment by jcribeirooo** [comment]
Hi @wviana, you need to setup the jdtls with the debug adapter plugin from Microsoft. There might be others, but I haven't bothered looking for it as this one works well.

[debug adapter](https://github.com/jcribeirooo/my-nvim-config/blob/master/lua/plugin/nvim-jdtls.lua#L18C1-L18C81)

You can get it from here and compile it yourself: https://github.com/microsoft/java-debug

**Comment by wviana** [comment]
@jcribeirooo, thanks for taking your time.

I have compiled and added to init_options bundles.

But I'm not executing this project using jdtls. I'm just running Java with those -agentlib parameters for debugging.

Jdtls is working fine. But I'm getting this handshake error when trying to connect with dad.

I wonder if this Java-debug adapter jar is actually being loaded. Or better, where should it be loaded? Together with my application or just having its path in the jdtls configuration should be enough?

**Comment by wviana** [comment]
I've taken a look at the jdtls code. The only place I've found using init_options.bundles was in dap.lua, but it all looks related to testing. I think that may run a new Java process. This is why I'm thinking that if the adapter shouldn't be in the same process, I'm trying to debug the class path.

**Comment by wviana** [comment]
@mfussenegger didn't want to bother you,  but any thoughts/tips on this?

**Comment by jcribeirooo** [comment]
Well, make sure you configured nvim jdtls correctly like in the guide. Refer to my config if you'd like to have some more lights.
You need to make sure nvim jdtls is the actual LSP you are using. Run :LspInfo and double check. I remember I had this problem where I didn't have the correct LSP attached to my buffer because of LSP Zero wrong configuration. Maybe it is your case.

**Comment by wviana** [comment]
Updates

So by finding in java-debug where it parses content-length and finding usages all the way up, I got into the executeCommand method in the JavaDebugDelegateCommandHandler class. I understand that somehow java-debug sits between JDI and DAP. I have no idea how. I even supposed that there was a way to execute something from java-debug saying what to execute, but I didn't find any main method in it.

**Comment by wviana** [comment]
By reading https://docs.oracle.com/en/java/javase/11/docs/specs/jpda/jpda_spis.html, I tried adding all those java-debug jars to the classpath when executing the application. I'm still getting the same error. 😢

**Comment by mfussenegger** [comment]
TL;DR: Work yourself carefully through the nvim-jdtls readme: https://github.com/mfussenegger/nvim-jdtls/ and you should be able to get this working.

---

Long version:


The JVM by default exposes debugging capabilities via the [java debug wire protocol (JDWP)](https://en.wikipedia.org/wiki/Java_Debug_Wire_Protocol).

If you run an application with `-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=5005 `, communication on port 5005 must happen using JDWP.

nvim-dap speaks the debug wire protocol, so you cannot connect directly to a java application with it:


```
                handshake error
  ┌──────────┐      ❌      ┌──────────────┐
  │ nvim-dap │   ────────►  │jdwp-agentlib │
  └──────────┘              └──────────────┘
```

This is where [java-debug](https://github.com/microsoft/java-debug) comes in. It is a _adapter_ from DAP to JDWP:


```

  ┌──────────┐    ✔️     ┌───────────┐   ✔️      ┌──────────────┐
  │ nvim-dap │  ──────► │java-debug │────────► │jdwp-agentlib │
  └──────────┘          └───────────┘          └──────────────┘
```


But java-debug isn't a regular application that you can execute directly in the shell or otherwise, but instead it is an extension to [eclipse.jdt.ls](https://github.com/eclipse-jdtls/eclipse.jdt.ls)

To use it you need to:

1. Make sure eclipse.jdt.ls loads it by adding the path to its jar to the `bundles` in the `initialization options` which LSP clients expose in some way.
2. Use a LSP command (`vscode.java.startDebugSession`) to start a debug session. The result to this command contains a port number
3. Have the debug adapter client connect to this port
4. Have the client do the initialization, and follow up with a `launch` or `attach` request.


nvim-jdtls is taking care of adding an adapter definition for nvim-dap, which covers 2) and 3).
It also contains functions for main-class discovery, which covers 4). 

Otherwise you can create attach configurations like in this `.vscode/launch.json`:

```
{
   "version": "0.2.0",
   "configurations": [
       {
           "type": "java",
           "request": "attach",
           "name": "Attach to hprofdap",
           "hostName": "127.0.0.1",
           "port": 5005
       }
   ]
}
```


How to do 1) is covered in the nvim-jdtls readme.
